Kernel/Net/Intel/E1000NetworkAdapter.cpp

```diff
@@ -247,202 +436,404 @@ UNMAP_AFTER_INIT E1000NetworkAdapter::~E1000NetworkAdapter() = default;
+    using E1000::has_flag;
 
-    m_entropy_source.add_random_event(status);
+    if (irq_cause == None)
+        return false;
 
-    if (status == 0)
+    // Let's be honest and only handle the interrupts we care about
+    if (!E1000::has_any_flag(irq_cause,
+            LSC | RXO | RXT0))
         return false;
 
-    if (status & INTERRUPT_LSC) {
-        u32 flags = in32(REG_CTRL);
-        out32(REG_CTRL, flags | ECTRL_SLU);
+    if (has_flag(irq_cause, LSC)) {
+        auto ctrl = m_registers.read<Register::Ctrl>();
+        ctrl.set_link_up = 1;
+        m_registers.write<Register::Ctrl>(ctrl);
+
+        m_link_up = m_registers.read<Register::Status>().link_up != 0;
 
-        m_link_up = ((in32(REG_STATUS) & STATUS_LU) != 0);
         autoconfigure_link_local_ipv6();
     }
-    if (status & INTERRUPT_RXDMT0) {
-        // Threshold OK?
-    }
-    if (status & INTERRUPT_RXO) {
+    if (has_flag(irq_cause, RXO)) {
         dbgln_if(E1000_DEBUG, "E1000: RX buffer overrun");
     }
-    if (status & INTERRUPT_RXT0) {
+    if (has_flag(irq_cause, RXT0)) {
+        // Note: RXDW on newer NICs, but it sounds like it has the same meaning
         receive();
     }
 
     m_wait_queue.wake_all();
 
-    out32(REG_INTERRUPT_CAUSE_READ, 0xffffffff);
+    m_registers.write<Register::InterruptCauseR>(InterruptClear);
     return true;
 }
 
 UNMAP_AFTER_INIT void E1000NetworkAdapter::detect_eeprom()
 {
-    out32(REG_EEPROM, 0x1);
+    if (m_operating_mode != OperatingMode::Intel8254x_legacy) {
+        // FIXME: Some models seem to lie here?
+        if (m_registers.read<Register::EEPROMControl>().eeprom_present)
+            m_has_eeprom.set();
+        else
+            dmesgln_pci(*this, "E1000: EEPROM not present");
+        return;
+    }
+
+    // The 82544GC/EI models do not have an EEPROM present bit
+    // So we cannot use that to determine if the EEPROM is present
+    // But have to try to read from it, to see if it's there
+    E1000::EEPROMRead eerd = {};
+    eerd.address_8.start = 1;
+    m_registers.write<Register::EEPROMRead>(eerd);
     for (int i = 0; i < 999; ++i) {
-        u32 data = in32(REG_EEPROM);
-        if (data & 0x10) {
+        auto data = m_registers.read<Register::EEPROMRead>();
+        if (data.address_8.done) {
             m_has_eeprom.set();
             return;
         }
+        Processor::wait_check();
     }
+
+    dmesgln_pci(*this, "E1000: EEPROM failed to initialize");
 }
 
-UNMAP_AFTER_INIT u32 E1000NetworkAdapter::read_eeprom(u8 address)
+UNMAP_AFTER_INIT u16 E1000NetworkAdapter::read_eeprom(u16 address)
 {
-    u16 data = 0;
-    u32 tmp = 0;
-    if (m_has_eeprom.was_set()) {
-        out32(REG_EEPROM, ((u32)address << 8) | 1);
-        while (!((tmp = in32(REG_EEPROM)) & (1 << 4)))
-            Processor::wait_check();
-    } else {
-        out32(REG_EEPROM, ((u32)address << 2) | 1);
-        while (!((tmp = in32(REG_EEPROM)) & (1 << 1)))
+    // FIXME: Should this just return 0 then?
```

Comment:
IMO yes. We could give an error back instead of crashing(?) the system, but that's personal preference.